File: solutions/270. Closest Binary Search Tree Value.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def closestValue(self, root: Optional[TreeNode], target: float) -> int:
        # soln 1 - time O(N), space O(N)
        ans: float('inf')
        diff = float('inf')
        while root:
            # ans = min(ans, root.val, key = lambda x: abs(target - x))
            val = abs(root.val-target)
            if val < diff or (val == diff and root.val < ans):
                ans = root.val
                diff = val
                
            if root.val > target:
                root = root.left
            elif root.val < target:
                root = root.right
            else:
                return root.val
        return ans
                

File: solutions/270. Closest Binary Search Tree Value.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    // Time O(N) (average O(logN)), space O(1)
public:
    void dfs(TreeNode* node, double target) {
        if (node == nullptr) 
            return;
        if (abs(node->val - target) < abs(target-ans_)) 
            ans_ = node->val;
        if (node->val > target)
            dfs(node->left, target);
        else
            dfs(node->right, target);
    }
    int closestValue(TreeNode* root, double target) {
        ans_ = root->val;
        dfs(root, target);
        return ans_;
    }
private:
    int ans_;
};
​
